core/src/graph/graph.rs

```diff
@@ -563,15 +615,82 @@ impl Graph {
 	/// Determine if page is full
 	///  aggressive:false -> use a simple heuristic based on the number of connections
 	///  aggressive:true  -> do actual compression to determine resulting actual page size
-	pub fn is_page_full(&self, page: &GraphPage, aggressive: bool) -> bool {
+	pub fn try_add_connection_to_page(
+		&self,
+		page: &mut GraphPage,
+		connection_id: &DsnpUserId,
+		aggressive: bool,
+		keys: &Option<ResolvedKeyPair>,
+		dsnp_version_config: Option<&DsnpVersionConfig>,
+	) -> Result<()> {
 		let connection_type = self.get_connection_type();
-		let max_connections_per_page = *PAGE_CAPACITIY_MAP.get(&connection_type).unwrap_or(&50);
+		let max_connections_per_page =
+			*PAGE_CAPACITY_MAP.get(&connection_type).unwrap_or_else(|| {
+				let mut capacities: Vec<&usize> = PAGE_CAPACITY_MAP.values().collect();
+				capacities.sort();
+				capacities.first().unwrap() // default: return smallest capacity value
+			});
 
 		if !aggressive {
-			return page.connections().len() >= max_connections_per_page
+			if page.connections().len() >= max_connections_per_page {
+				return Err(Error::msg("Page trivially full"))
+			}
+
+			if let Err(e) = page.add_connection(connection_id) {
+				return Err(e)
+			}
+
+			return Ok(())
+		}
+
+		if connection_type.privacy_type() == PrivacyType::Private &&
+			(dsnp_version_config.is_none() || keys.is_none())
+		{
+			return Err(Error::msg("Missing encryption key or config"))
+		}
+
+		let max_page_size = self.environment.get_config().max_graph_page_size_bytes as usize;
+		// Max page fullness leaves room for 2 uncompressed connections
+		let max_page_fullness = max_page_size -
+			(2 as usize *
+				(std::mem::size_of::<DsnpGraphEdge>() + std::mem::size_of::<DsnpPrid>()));
+
+		let mut temp_page = page.clone();
+		if let Err(e) = temp_page.add_connection(connection_id) {
+			return Err(e)
 		}
 
-		todo!()
+		let page_blob = match connection_type {
+			ConnectionType::Follow(PrivacyType::Public) |
+			ConnectionType::Friendship(PrivacyType::Public) => temp_page.to_public_page_data(),
+			ConnectionType::Follow(PrivacyType::Private) => {
+				let keys = keys.as_ref().unwrap();
+				let dsnp_version_config = dsnp_version_config.unwrap();
+				temp_page.clear_prids();
+				temp_page.to_private_page_data(dsnp_version_config, &keys)
+			},
+			ConnectionType::Friendship(PrivacyType::Private) => {
+				let keys = keys.as_ref().unwrap();
+				let dsnp_version_config = dsnp_version_config.unwrap();
+				self.apply_prids(&mut temp_page, &vec![], &keys)
+					.expect("Error applying prids to page");
+				temp_page.to_private_page_data(dsnp_version_config, &keys)
+			},
+		};
+
+		match page_blob {
+			Ok(blob) =>
+				if blob.content.len() >= max_page_fullness {
```

Comment:
is this `>` or `>=` ?

Reply:
Not that important, since it only makes a difference of 1 byte, and since the size of the "safety padding" is larger than that, it's not a big deal.

But, you're right--it's `>`. Fixed. 😬